nengo_gui/server.py

```diff
@@ -1,171 +1,519 @@
-                             '<a href="#" rel="%s">%s</a></li>' % (e,ff,f))
-        r.append('</ul>')
-        return ''.join(r)
-
-    def swi_static(self, *path):
-        """Handles http://host:port/static/* by returning pkg data"""
-        fn = os.path.join('static', *path)
-        mimetype, encoding = mimetypes.guess_type(fn)
-        data = pkgutil.get_data('nengo_gui', fn)
-        return (mimetype, data)
-
-    def swi_favicon_ico(self):
-        icon = pkgutil.get_data('nengo_gui', 'static/favicon.ico')
-        return ('image/ico', icon)
-
-    def create_login_form(self):
-        if self.attempted_login:
-            message = 'Invalid password. Try again.'
+                self.http_GET()
+        except HttpError as err:
+            logger.warning(
+                'Error response (%i): %s', err.code, err.msg, exc_info=True)
+            err.to_response().send(self)
+        except Exception as err:
+            logger.exception('Error response')
+            err = InternalServerError(
+                '<pre>' + traceback.format_exc() + '</pre>')
+            err.to_response().send(self)
+
+    def http_GET(self):
+        command = self._get_command(self.http_commands, self.resource)
+        if command is None:
+            response = self.http_default()
         else:
-            message = 'Enter the password:'
-        return """<form action="/" method=GET>%s<br>
-            <input type=hidden name=swi_id value=''>
-            <input type=password name=swi_pwd>
-            <input type=submit value="Log In">
-            </form>""" % message
+            response = getattr(self, command)()
+        response.send(self)
+
+    def http_default(self):
+        raise InvalidResource(self.path)
+
+    def handle_upgrade(self):
+        upgrade = self.headers.get('Upgrade').lower()
+        if upgrade == 'websocket':
+            self.upgrade_to_ws()
+        else:
+            raise BadRequest()
+
+    def upgrade_to_ws(self):
+        response = '''HTTP/1.1 101 Switching Protocols
+Upgrade: websocket
+Connection: Upgrade
+Sec-WebSocket-Accept: {sec}
 
-    def swi(self, filename=None, reset=None):
-        """Handles http://host:port/ by giving the main page"""
-        if self.user is None:
-            return self.create_login_form()
+'''
+        valid_srv_addrs = [
+            '{0}:{1}'.format(
+                self.server.server_name.lower(), self.server.server_port),
+            '{0}:{1}'.format(socket.getfqdn(), self.server.server_port)]
+        if self.server.server_port == 80:
+            valid_srv_addrs.append(self.server.settings.listen_addr[0].lower())
+            valid_srv_addrs.append(socket.getfqdn())
 
-        reset_cfg = reset == 'True'
+        try:
+            origin = urlparse(self.headers['Origin'])
+            assert origin.netloc.lower() in valid_srv_addrs
+        except KeyError:
+            raise Forbidden()
+        except AssertionError:
+            raise Forbidden()
```

Comment:
On windows, this still raises an exception for me.  When it happens, ```origin.netloc``` is ```localhost:8080``` and ```valid_srv_addrs``` is ```['amygdala:8080', 'amygdala:8080']```    ("amygdala" is my computer name)